### DISCUSSION TITLE
Weighted mean

### DISCUSSION THREAD

**Discussion by rsneumann** [jtablesaw/tablesaw#1049] [Q&A]
Hi, I would like to know how to calculate the weighted mean with Tablesaw. For instance, for the following table:

Name ItemCost Quantity
A 13 3
A 7 2
B 9 2

I would like the result

Name Weigthedmean
A 10.6
B 9

Of course, other weighted statistics (such as Std Dev etc) are also of interest!
Thanks for any suggestion!

**Comment by lwhite1** [comment]
weirdly, i don't think i've ever done this. It should probably be added as a function. 

Here's some code that does it. I put in a bunch of println statements so you can see the results being built.

```
    public static void main(String[] args) {
        Table t = Table.read().csv(
                """
                        Name,cost,Quantity
                        A,13,3
                        A,7,2
                        B,9,2
                        """, "test");
        System.out.println(t);
        t.addColumns(t.intColumn("cost").multiply(t.intColumn("quantity")));
        System.out.println(t);
        Table sums = t.summarize("quantity", "cost * quantity", sum).by("name");
        System.out.println(sums);
        sums.addColumns(sums.doubleColumn("Sum [cost * Quantity]").divide(sums.doubleColumn("sum [Quantity]")));
        System.out.println(sums);
    }

```

And here's the output:

```
             test             
 Name  |  cost  |  Quantity  |
------------------------------
    A  |    13  |         3  |
    A  |     7  |         2  |
    B  |     9  |         2  |
                       test                       
 Name  |  cost  |  Quantity  |  cost * Quantity  |
--------------------------------------------------
    A  |    13  |         3  |               39  |
    A  |     7  |         2  |               14  |
    B  |     9  |         2  |               18  |
                    test summary                     
 Name  |  Sum [cost * Quantity]  |  Sum [Quantity]  |
-----------------------------------------------------
    A  |                     53  |               5  |
    B  |                     18  |               2  |
                                          test summary                                          
 Name  |  Sum [cost * Quantity]  |  Sum [Quantity]  |  Sum [cost * Quantity] / Sum [Quantity]  |
------------------------------------------------------------------------------------------------
    A  |                     53  |               5  |                                    10.6  |
    B  |                     18  |               2  |                                       9  |


```

Note that any columnwise math returns results as DoubleColumn. 
cheers

**Reply by rsneumann** [reply]
Thanks for your answer! Doing it "manually" as you have shown is possible, of course. But I was hoping it would be possible to come up with a function for that, something like 
tbl.summarize("ItemCost", "Quantity", weightedMeanFunc).by("Name")
I have tried implementing such a function myself, inheriting from "NumericAggregateFunction". But the problem is that here we have to combine two columns into one final aggregated value. So I guess there is no creative way of creating such a function inside the existing framework? I have to repeatedly calculate weigthed means, standard deviation etc, so the above solution becomes a bit verbose... 
Thanks again!

**Comment by lwhite1** [comment]
it's probably possible, but not on the existing summarization framework.
